### DISCUSSION TITLE
CUDA issue with local RootPainter server WIN

### DISCUSSION THREAD

**Discussion by lamewarden** [Abe404/root_painter#113] [Q&A] [answered]
Hi. I was setting local RP server on windows machine with Nvidia RTX 4090 with installed compatible CUDA toolkit, NVIDIA drivers, pytotch from https://pytorch.org/get-started/locally/ and all required packages from requirements.txt.
However still get errors while trying to segment something with pre-trained model.
It doesn't detect GPU for some reason:
![выява](https://github.com/Abe404/root_painter/assets/8532070/c2966cef-98fc-4944-a9e7-85f39334215b)

Here is the text of error hwich I get when it tries to read instructions and start segmentation:
```
Traceback (most recent call last):
  File "D:\root_tracker\body\trainer\model_utils.py", line 42, in load_model
    model.load_state_dict(torch.load(model_path))
                          ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gpt\AppData\Local\Programs\Python\Python311\Lib\site-packages\torch\serialization.py", line 809, in load
    return _load(opened_zipfile, map_location, pickle_module, **pickle_load_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gpt\AppData\Local\Programs\Python\Python311\Lib\site-packages\torch\serialization.py", line 1172, in _load
    result = unpickler.load()
             ^^^^^^^^^^^^^^^^
  File "C:\Users\gpt\AppData\Local\Programs\Python\Python311\Lib\site-packages\torch\serialization.py", line 1142, in persistent_load
    typed_storage = load_tensor(dtype, nbytes, key, _maybe_decode_ascii(location))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gpt\AppData\Local\Programs\Python\Python311\Lib\site-packages\torch\serialization.py", line 1116, in load_tensor
    wrap_storage=restore_location(storage, location),
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gpt\AppData\Local\Programs\Python\Python311\Lib\site-packages\torch\serialization.py", line 217, in default_restore_location
    result = fn(storage, location)
             ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gpt\AppData\Local\Programs\Python\Python311\Lib\site-packages\torch\serialization.py", line 182, in _cuda_deserialize
    device = validate_cuda_device(location)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gpt\AppData\Local\Programs\Python\Python311\Lib\site-packages\torch\serialization.py", line 166, in validate_cuda_device
    raise RuntimeError('Attempting to deserialize object on a CUDA '
RuntimeError: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.
```

Could you please give me a hint what can be the issue?

**Comment by Abe404** [comment]
Hi!

thanks for reaching out!

My best guess is that cuda (or PyTorch with cuda support) is not yet installed properly on your system. It can be a little tricky.

See the following stack overflow answer, which includes a code snippet showing how to check if your PyTorch installation has access to your GPU.

https://stackoverflow.com/a/48152675

The snippet included inline for convenience:

‘’’
>>> import torch
>>> torch.cuda.is_available()
True
‘’’

The cuda and PyTorch version you need to install is specific to your platform. I suggest figuring this bit out before installing the RootPainter server, as the RootPainter server should work with any recent version of PyTorch. 

Unfortunately there’s a lot that could have gone wrong so I can’t really give you too much advice on installing cuda and PyTorch on your platform beyond following the official NVIDIA and PyTorch documentation. I don’t believe this issue is within RootPainter itself (although I appreciate the cuda/PyTorch dependency can be a headache to sort out and I wish I knew of a better solution).

The NVIDIA 4090 GPU is an excellent choice for running RootPainter. I am sure you will have fun once you get cuda and PyTorch installed.

Kind regards,
Abraham

**Comment by Abe404** [comment]
A related discussion: https://discuss.pytorch.org/t/windows-11-with-nvidia-rtx-4090/170221/13


Perhaps this could help

**Comment by Abe404** [comment]
Another option could be to try the nightly PyTorch and see if that gives you GPU access:
pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu118
(from: https://pytorch.org/get-started/locally/) 
The URL should be modified based on your cuda version.

**Reply by lamewarden** [reply, marked as answer]
Thank you for all the comments! All problems were solved when I installed NVIDIA CUDA with Anaconda. With pytorch nightly for cuda 12.1. It works perfectly now.

**Reply by Abe404** [reply]
Thanks for the update. I am really happy you were able to figure it out.

**Reply by lamewarden** [reply]
Hi Abraham, is there a way to contact with you privately? I've got some collaboration offer.

**Reply by Abe404** [reply]
Exciting stuff! 

My university email is visible as the corresponding author on the paper: https://nph.onlinelibrary.wiley.com/doi/full/10.1111/nph.18387

click my name in the list of authors on the webpage and it shows up.

Otherwise it’s possible to DM me on twitter if you use that:

https://twitter.com/Abe_404

kind regards,
Abraham
